src/index.js

```diff
@@ -1,98 +1,140 @@
+    dampingDrop = 0.1,
+    dampingBoost = 1.5,
+    maxDamping = Number.MAX_SAFE_INTEGER,
+    minDamping = Number.EPSILON,
     maxValues,
-    initialValues
-  } = options;
+    minValues,
+    residualEpsilon = 1e-6,
+    initialValues,
+    errorPropagation = 50
+  } = legacy.compatOptions(options);
 
   if (damping <= 0) {
     throw new Error('The damping option must be a positive number');
-  } else if (!data.x || !data.y) {
-    throw new Error('The data parameter must have x and y elements');
+  } else if (!data || !data.x || !data.y) {
+    throw new Error('The data object must have x and y elements');
   } else if (
     !Array.isArray(data.x) ||
     data.x.length < 2 ||
     !Array.isArray(data.y) ||
     data.y.length < 2
   ) {
+    throw new Error('The data must have more than 2 points');
+  } else if (data.x.length !== data.y.length) {
     throw new Error(
-      'The data parameter elements must be an array with more than 2 points'
+      'The data object must have equal number of x and y coordinates'
     );
-  } else if (data.x.length !== data.y.length) {
-    throw new Error('The data parameter elements must have the same size');
   }
 
-  var parameters =
-    initialValues || new Array(parameterizedFunction.length).fill(1);
-  let parLen = parameters.length;
+  let params = initialValues || new Array(paramFunction.length).fill(1);
+  let parLen = params.length;
   maxValues = maxValues || new Array(parLen).fill(Number.MAX_SAFE_INTEGER);
   minValues = minValues || new Array(parLen).fill(Number.MIN_SAFE_INTEGER);
 
-  if (maxValues.length !== minValues.length) {
-    throw new Error('minValues and maxValues should have the same size');
+  if (!Array.isArray(params)) {
+    throw new Error('initialValues must be an array');
   }
 
-  if (!Array.isArray(parameters)) {
-    throw new Error('initialValues must be an array');
+  if (
+    maxValues.length !== minValues.length ||
+    maxValues.length !== params.length
+  ) {
+    throw new Error(
+      'minValues and maxValues must have the same size as the number of parameters'
+    );
   }
 
-  var error = errorCalculation(data, parameters, parameterizedFunction);
+  initializeErrorPropagation(errorPropagation);
+
+  /** @type Array<number> */
+  var residualDifferences = Array(10).fill(NaN);
 
-  var converged = error <= errorTolerance;
+  var residuals = sumOfSquaredResiduals(data, params, paramFunction);
+  var converged = false;
 
   for (
     var iteration = 0;
     iteration < maxIterations && !converged;
     iteration++
   ) {
-    parameters = step(
+    var params2 = step(
       data,
-      parameters,
+      params,
       damping,
       gradientDifference,
-      parameterizedFunction
+      paramFunction
     );
 
     for (let k = 0; k < parLen; k++) {
-      parameters[k] = Math.min(
-        Math.max(minValues[k], parameters[k]),
-        maxValues[k]
-      );
+      params2[k] = Math.min(Math.max(minValues[k], params2[k]), maxValues[k]);
```

Comment:
`params2` seems like a name that isn't highly expressive. Am I understanding correctly that what you're doing here is avoiding overwriting the previous parameters in case they produce better results? If so, perhaps it'd be better to use names like `prevParams` and `updatedParams` or `lastParams` and `nextParams`, etc.